### ISSUE TITLE
Crash in initTZ when multiple connections are used.

### ISSUE THREAD

**Issue by kurt-google** [godror/godror#249]
**Describe the bug**
I am observing a crash in `initTZ` when multiple connections are used. It appears to be due to unsafe sharing of `Driver`state between connectors.

Specifically I see `panic: assignment to entry in nil map` pointed at https://github.com/godror/godror/blob/main/conn.go#L602 during connection initialization.

This appears due to the fact that we create multiple connections within the same process resulting in multiple `connector`s that can be `Closed` via https://github.com/godror/godror/blob/main/drv.go#L1127 so when one connection is closed (`(*sql.DB).Close()`) while we are trying to establish a 2nd connection with `sql.Open("godror", ...)` a race can occur and result in this panic.

Due to nil maps still being "readable" this doesnt cause any problems with the initial reads, only once the map is written do we see the panic.

**To Reproduce**
Opening and closing many parallel `*sql.DB` should reproduce this.

**Expected behavior**
It seems  that this timezone information could be pushed down to the individual `connector` which corresponds to the `Close()` target that is cleaning it up.

**Error output**
You'll have to forgive that I'm still using an ancient version (0.25.3) but this still looks applicable and we are trying to diagnose other issues blocking the upgrade to later releases.
```
github.com/godror/godror.(*conn).initTZ(0xc0000d6240)
	__main__/external/com_github_godror_godror/conn.go:567 +0xfe9
github.com/godror/godror.(*conn).init(0xc0000d6240, {0xe7ea40, 0xc0005de660}, 0x0)
	__main__/external/com_github_godror_godror/conn.go:463 +0x10e
github.com/godror/godror.(*drv).createConn(_, _, {{{0xc0002cc4a0, 0xe}, {0xc000036330, 0x15}, {{0xc0002cc4c0, 0xa}}, {0x0, 0x0}, ...}, ...})
	__main__/external/com_github_godror_godror/drv.go:392 +0x355
github.com/godror/godror.(*drv).createConnFromParams(_, {{{0xc0002cc4a0, 0xe}, {0xc000036330, 0x15}, {{0xc0002cc4c0, 0xa}}, {0x0, 0x0}, {0x0, ...}, ...}, ...})
	__main__/external/com_github_godror_godror/drv.go:587 +0x267
github.com/godror/godror.connector.Connect({0x13c7dc0, {{{0xc0002cc4a0, 0xe}, {0xc000036330, 0x15}, {{...}}, {0x0, 0x0}, {0x0, 0x0}, ...}, ...}}, ...)
	__main__/external/com_github_godror_godror/drv.go:1087 +0x798
database/sql.(*DB).conn(0xc0003ac9c0, {0xe7ea78, 0xc0005161e0}, 0x1)
	GOROOT/src/database/sql/sql.go:1395 +0x782
database/sql.(*DB).PingContext(0xd47fd2?, {0xe7ea78, 0xc0005161e0})
	GOROOT/src/database/sql/sql.go:853 +0x79
github.com/GoogleCloudPlatform/elcarro-oracle-operator/oracle/pkg/database/dbdaemon.(*Server).CheckDatabaseState(0xc00049c000, {0xe7ea78, 0xc0005161e0}, 0xc000524320)
	oracle/pkg/database/dbdaemon/dbdaemon_server.go:1057 +0x578
```

**Your oracle client version**
19.16

**Your godror version**
v0.25.3

**Your go version**
v1.18.5

**Your gcc version**
11.3.0

**Machine (please complete the following information):**
 - OS: Linux
 - Architecture: 86_64
 - Version: EL 8

**Additional context**

**Comment by tgulacsi** [comment]
```git log --oneline v0.25.3..HEAD drv.go``` lists 30 commits, a few directly messing with drv/conn. So maybe this is already fixed.

I've added TestOpenCloseConn to z_test.go, but can't reproduce it neither on current master, nor on v0.25.3 - so the test does not test what you see.
Can you make the test fail ? It's on a new issue249 branch.
I've also added a make for c.drv.timezones, but it's fishy - it shouldn't be nil.

A proper (failing) test would help diagnose the real problem.

**Comment by kurt-google** [comment]
I've added a replicator on top of `issue249` branch in my repository https://github.com/kurt-google/godror/commit/ba3b8a7e95022896c9f634005883e0e8cbd68b5d

This replicates 4/5 times on my machine.

**Comment by tgulacsi** [comment]
I cannot reproduce it - neither on the Autonomous DB[^1], nor on our test DB[^2] .

I've tested it with ```go test -run=Crash -count=10```

---

[^1]:  standalone connections, 
    Version: v0.25.3                               
    Client: 21.4.0.0.0 Timezone: Local         
    Server: 19.16.0.1.0 [Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production; Version 19.16.0.1.0] Timezone: UTC


[^2]: pooled connections,
    Version: v0.25.3
    Client: 21.4.0.0.0 Timezone: Local
    Server: 19.16.0.0.0 [Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production; Version 19.16.0.0.0] Timezone: Local

**Comment by tgulacsi** [comment]
```go test -run=Crash -count=10 -race``` surfaced a race condition with pooled connections on v0.25.3, but not with the current main (issue249-main).

**Comment by kurt-google** [comment]
The atomics used in the modification of my replicator end up widening the race between threads so it was more likely that both threads waited, using a mutex would probably be easier to solve the race. But I was still able to replicate it maybe 25/50% of the time with that implementation.

However I rebased my branch onto v0.34 and it seems i couldnt trigger the issue anymore. Ill take another look at the conn changes in 34.

All my tests are done with a local 19.16 server and server client libs. For reference on testing process I compile the test with `go test -c z_249_test.go` and then run it directly `./godror.test -test.v`

**Comment by kurt-google** [comment]
It looks like there was yet another race in my test code, the deferred close should instead have occurred before the join to ensure the db was actually closed. I have updated the issue249 branch on my repository. It should now reproduce more consistently on v0.25.3.

Trying to bisect this, it seems that at `dbc4bb2` the issue stops appearing, in particular the db.Close() never propagates to the driver after this point. Prior to this the crash replicates in the main tree (a620931) and the timeout tree (505349a) causes the test to deadlock.

**Comment by tgulacsi** [comment]
Sorry, I cannot reproduce it.

**Comment by kurt-google** [comment]
Thanks for taking a look, its a bit concerning you cannot get it to reproduce with my code but since it seems close is never called in v0.34 closing seems fine. I also confirmed the deadlock from the timeout tree was fixed in https://github.com/godror/godror/commit/dbc4bb230d602eb443cc66403a06967b46e7daac
